File: internal/fsx/fsx.go
// SPDX-License-Identifier: Apache-2.0
// Atomic filesystem primitives. mkdir, open(O_EXCL), rename. Nothing else.
// Ported from src/core/fsx.mjs: every mutable record is staged in tmp/, fsynced
// and renamed, so a reader sees the whole old version or the whole new one.
package fsx

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/RagnarPitla/agent-fridge/internal/errs"
	"github.com/RagnarPitla/agent-fridge/internal/jsonx"
	"github.com/RagnarPitla/agent-fridge/internal/util"
)

var beforeExclusivePublish = func(_, _ string) {}

const isWindows = runtime.GOOS == "windows"

// EnsureDir creates a directory tree, mapping a read-only or forbidden target
// onto E_PERMISSION rather than a raw errno.
func EnsureDir(dir string) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		if errors.Is(err, os.ErrPermission) {
			return errs.New("E_PERMISSION", "Cannot create "+dir+": "+err.Error())
		}
		return err
	}
	return nil
}

func fsyncDir(dir string) {
	if isWindows {
		return
	}
	f, err := os.Open(dir)
	if err != nil {
		return
	}
	_ = f.Sync()
	_ = f.Close()
}

// WriteAtomic stages bytes in tmpDir, fsyncs them, and renames over the target.
// The rename is retried on Windows, where an antivirus scanner or the indexer
// can hold a transient handle and turn a legal replace into EPERM.
func WriteAtomic(finalPath, text, tmpDir string) error {
	dir := filepath.Dir(finalPath)
	if err := EnsureDir(dir); err != nil {
		return err
	}
	if err := EnsureDir(tmpDir); err != nil {
		return err
	}
	tmp := filepath.Join(tmpDir, strconv.Itoa(os.Getpid())+"-"+util.ULID()+".tmp")
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o600)
	if err != nil {
		if errors.Is(err, os.ErrPermission) {
			return errs.New("E_PERMISSION", "Cannot write "+finalPath+": "+err.Error())
		}
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	if _, err := f.WriteString(text); err != nil {
		_ = f.Close()
		_ = os.Remove(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	if err := f.Sync(); err != nil {
		_ = f.Close()
		_ = os.Remove(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	if err := f.Close(); err != nil {
		_ = os.Remove(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	renamed := false
	for attempt := 1; attempt <= 6 && !renamed; attempt++ {
		err = os.Rename(tmp, finalPath)
		if err == nil {
			renamed = true
			break
		}
		if isWindows && attempt < 6 {
			time.Sleep(time.Duration(10*(1<<(attempt-1))) * time.Millisecond)
			continue
		}
		_ = os.Remove(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed replacing "+finalPath+": "+err.Error())
	}
	fsyncDir(dir)
	return nil
}

// CreateExclusive is the write-once path used for notes: O_EXCL means two
// processes can never end up writing the same file.
func CreateExclusive(finalPath, text string, tmpDir string) error {
	dir := filepath.Dir(finalPath)
	if err := EnsureDir(dir); err != nil {
		return err
	}
	staging := tmpDir
	if staging == "" {
		staging = filepath.Join(dir, ".tmp")
	}
	if err := EnsureDir(staging); err != nil {
		return err
	}
	tmp := filepath.Join(staging, fmt.Sprintf("%d-%s.tmp", os.Getpid(), util.ULID()))
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
	if err != nil {
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	if _, err := f.WriteString(text); err != nil {
		_ = f.Close()
		UnlinkQuiet(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	if err := f.Sync(); err != nil {
		_ = f.Close()
		UnlinkQuiet(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	if err := f.Close(); err != nil {
		UnlinkQuiet(tmp)
		return errs.New("E_STATE_CORRUPT", "Failed staging "+finalPath+": "+err.Error())
	}
	beforeExclusivePublish(tmp, finalPath)
	// Link, not open+write. A reader that opens the final name always sees the
	// whole note, because the name only exists once the bytes are on disk.
	if err := os.Link(tmp, finalPath); err != nil {
		if errors.Is(err, fs.ErrExist) {
			UnlinkQuiet(tmp)
			return err
		}
		// Filesystems without hard links (some SMB shares, FAT). Rename is
		// still atomic in content, and the caller's names carry a random id,
		// so losing link's exclusivity here costs nothing that matters.
		if Exists(finalPath) {
			UnlinkQuiet(tmp)
			return fs.ErrExist
		}
		if err := os.Rename(tmp, finalPath); err != nil {
			UnlinkQuiet(tmp)
			return err
		}
		return nil
	}
	UnlinkQuiet(tmp)
	return nil
}

// WriteJSONAtomic writes a record in the protocol's canonical JSON form.
func WriteJSONAtomic(path string, obj jsonx.Obj, tmpDir string) error {
	return WriteAtomic(path, jsonx.Stable(obj), tmpDir)
}

// CreateJSONExclusive writes a write-once record in canonical JSON form.
func CreateJSONExclusive(path string, obj jsonx.Obj, tmpDir string) error {
	return CreateExclusive(path, jsonx.Stable(obj), tmpDir)
}

// ReadJSON reads a record, reporting unparseable bytes as E_STATE_CORRUPT
// rather than crashing or silently skipping.
func ReadJSON(file string) (jsonx.Obj, error) {
	raw, err := os.ReadFile(file)
	if err != nil {
		return nil, err
	}
	obj, err := jsonx.ParseObj(raw)
	if err != nil {
		return nil, errs.New("E_STATE_CORRUPT", "Unparseable record: "+file).WithHint("fridge doctor --fix")
	}
	return obj, nil
}

// ReadJSONSafe never fails: a damaged record becomes a finding, not a crash.
func ReadJSONSafe(file string) (jsonx.Obj, bool) {
	obj, err := ReadJSON(file)
	if err != nil {
		return nil, false
	}
	return obj, true
}

// ListJSON returns the .json files directly inside dir, sorted by filename, in
// the same order Array#sort gives the Node implementation.
func ListJSON(dir string) []string {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return []string{}
	}
	names := []string{}
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".json") {
			names = append(names, e.Name())
		}
	}
	sort.Slice(names, func(i, j int) bool { return jsonx.LessUTF16(names[i], names[j]) })
	out := make([]string, 0, len(names))
	for _, n := range names {
		out = append(out, filepath.Join(dir, n))
	}
	return out
}

// WalkJSON returns every .json file under dir, depth first, name sorted.
func WalkJSON(dir string) []string {
	out := []string{}
	walkJSON(dir, &out)
	return out
}

func walkJSON(dir string, out *[]string) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return
	}
	names := make([]string, 0, len(entries))
	isDir := map[string]bool{}
	for _, e := range entries {
		names = append(names, e.Name())
		isDir[e.Name()] = e.IsDir()
	}
	sort.Slice(names, func(i, j int) bool { return jsonx.LessUTF16(names[i], names[j]) })
	for _, n := range names {
		full := filepath.Join(dir, n)
		if isDir[n] {
			walkJSON(full, out)
		} else if strings.HasSuffix(n, ".json") {
			*out = append(*out, full)
		}
	}
}

// Exists reports whether a path is there at all.
func Exists(p string) bool {
	_, err := os.Lstat(p)
	return err == nil
}

// IsDir reports whether a path is a directory, following symlinks.
func IsDir(p string) bool {
	st, err := os.Stat(p)
	return err == nil && st.IsDir()
}

// RmRF removes a tree and does not complain when it was already gone.
func RmRF(p string) { _ = os.RemoveAll(p) }

// UnlinkQuiet removes a file and treats a missing file as success.
func UnlinkQuiet(p string) { _ = os.Remove(p) }

// ReadTextOr reads a file, returning ok=false when it cannot be read at all.
func ReadTextOr(p string) (string, bool) {
	b, err := os.ReadFile(p)
	if err != nil {
		return "", false
	}
	return string(b), true
}

File: src/core/fsx.mjs
// SPDX-License-Identifier: Apache-2.0
// Atomic filesystem primitives. mkdir, open(O_EXCL), rename. Nothing else.
import fs from 'node:fs';
import path from 'node:path';
import { AppError } from './errors.mjs';
import { stableStringify, ulid } from './util.mjs';

const IS_WIN = process.platform === 'win32';
const RENAME_RETRY = ['EPERM', 'EACCES', 'EBUSY'];

export function ensureDir(dir) {
  try {
    fs.mkdirSync(dir, { recursive: true });
  } catch (e) {
    if (e.code === 'EACCES' || e.code === 'EROFS') throw new AppError('E_PERMISSION', `Cannot create ${dir}: ${e.code}`);
    throw e;
  }
}

function sleepSync(ms) {
  const end = Date.now() + ms;
  while (Date.now() < end) {
    /* deliberate spin: sync path, sub-350ms worst case */
  }
}

function fsyncDir(dir) {
  if (IS_WIN) return;
  let fd;
  try {
    fd = fs.openSync(dir, 'r');
    fs.fsyncSync(fd);
  } catch {
    /* directory fsync is best effort */
  } finally {
    if (fd !== undefined) try { fs.closeSync(fd); } catch { /* ignore */ }
  }
}

/** Stage in tmp, fsync, rename. The only way a mutable record is ever written. */
export function writeAtomic(finalPath, text, tmpDir) {
  const dir = path.dirname(finalPath);
  ensureDir(dir);
  ensureDir(tmpDir);
  const tmp = path.join(tmpDir, `${process.pid}-${ulid()}.tmp`);
  let fd;
  try {
    fd = fs.openSync(tmp, 'wx', 0o600);
    fs.writeFileSync(fd, text);
    fs.fsyncSync(fd);
  } catch (e) {
    if (fd !== undefined) try { fs.closeSync(fd); } catch { /* ignore */ }
    try { fs.unlinkSync(tmp); } catch { /* ignore */ }
    if (e.code === 'EACCES' || e.code === 'EROFS') throw new AppError('E_PERMISSION', `Cannot write ${finalPath}: ${e.code}`);
    throw new AppError('E_STATE_CORRUPT', `Failed staging ${finalPath}: ${e.message}`);
  } finally {
    if (fd !== undefined) try { fs.closeSync(fd); } catch { /* ignore */ }
  }
  let renamed = false;
  for (let attempt = 1; attempt <= 6 && !renamed; attempt++) {
    try {
      fs.renameSync(tmp, finalPath);
      renamed = true;
    } catch (e) {
      if (IS_WIN && RENAME_RETRY.includes(e.code) && attempt < 6) {
        sleepSync(10 * 2 ** (attempt - 1));
        continue;
      }
      try { fs.unlinkSync(tmp); } catch { /* ignore */ }
      throw new AppError('E_STATE_CORRUPT', `Failed replacing ${finalPath}: ${e.code || e.message}`);
    }
  }
  fsyncDir(dir);
  return finalPath;
}

/** Write-once create. Used for notes: a note is written exactly once, never rewritten. */
export function createExclusive(finalPath, text) {
  ensureDir(path.dirname(finalPath));
  const fd = fs.openSync(finalPath, 'wx', 0o644);
  try {
    fs.writeFileSync(fd, text);
    fs.fsyncSync(fd);
  } finally {
    fs.closeSync(fd);
  }
  return finalPath;
}

export const writeJsonAtomic = (p, obj, tmpDir) => writeAtomic(p, stableStringify(obj), tmpDir);
export const createJsonExclusive = (p, obj) => createExclusive(p, stableStringify(obj));

export function readJson(file) {
  const raw = fs.readFileSync(file, 'utf8');
  try {
    return JSON.parse(raw);
  } catch (e) {
    throw new AppError('E_STATE_CORRUPT', `Unparseable record: ${file}`, { hint: 'fridge doctor --fix' });
  }
}

export function readJsonSafe(file) {
  try {
    return { ok: true, value: readJson(file) };
  } catch (e) {
    return { ok: false, error: e.message, file };
  }
}

export function listJson(dir) {
  try {
    return fs.readdirSync(dir).filter((f) => f.endsWith('.json')).sort().map((f) => path.join(dir, f));
  } catch (e) {
    if (e.code === 'ENOENT') return [];
    throw e;
  }
}

export function walkJson(dir, out = []) {
  let entries;
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch (e) {
    if (e.code === 'ENOENT') return out;
    throw e;
  }
  for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) walkJson(full, out);
    else if (entry.name.endsWith('.json')) out.push(full);
  }
  return out;
}

export const exists = (p) => fs.existsSync(p);
export const rmrf = (p) => fs.rmSync(p, { recursive: true, force: true });
export const unlinkQuiet = (p) => { try { fs.unlinkSync(p); } catch { /* ignore */ } };
